compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -0,0 +1,812 @@
+            _uscgChoice
+          chupds = concatMap (\(self,this,args,upds) -> _ussUpdate $ upds (EVar self) (EVar this) (EVar args)) lookupRes
+      in (exp, UpdateSetS (_uscgUpdate ++ chupds))
+
+    convertChHMap :: HM.HashMap UpdChoice (ExprVarName,ExprVarName,ExprVarName,Expr -> Expr -> Expr -> UpdateSet 'ChoiceGathering)
+      -> HM.HashMap UpdChoice (ExprVarName,ExprVarName,ExprVarName,Expr -> Expr -> Expr -> UpdateSet 'Solving)
+    convertChHMap = HM.map (\(self,this,args,updfunc) ->
+      (self,this,args, \(selfExp :: Expr) (thisExp :: Expr) (argsExp :: Expr) ->
+        updateSetCG2S $ updfunc selfExp thisExp argsExp))
+
+    updateSetCG2S :: UpdateSet 'ChoiceGathering -> UpdateSet 'Solving
+    updateSetCG2S UpdateSetCG{..} = if null _uscgChoice
+      then UpdateSetS _uscgUpdate
+      else error "Impossible: There should be no references remaining after choice solving"
+
+-- | Solves a single reference by recursively inlining the references into updates.
+solveReference :: forall updset ref. (Eq ref, Hashable ref)
+  => (ref -> HM.HashMap ref updset -> updset)
+  -- ^ Function for looking up references in the update set.
+  -> (updset -> ([Cond ref], updset))
+  -- ^ Function popping the references from the update set.
+  -> (updset -> updset -> updset)
+  -- ^ Function for concatinating update sets.
+  -> (Cond updset -> updset)
+  -- ^ Function for moving conditionals inside the update set.
+  -> [ref]
+  -- ^ The references which have already been visited.
+  -> HM.HashMap ref updset
+  -- ^ The hashmap mapping references to update sets.
+  -> ref
+  -- ^ The reference to be solved.
+  -> (updset, HM.HashMap ref updset)
+solveReference lookup getRefs extUpds introCond vis hmap0 ref0 =
+  let upd0 = lookup ref0 hmap0
+      (refs, upd1) = getRefs upd0
+  in if ref0 `elem` vis
+  -- TODO: Recursion!
+    then (upd1, hmap0) -- TODO: At least remove the references?
+    else let (upd2, hmap1) = foldl handle_ref (upd1, hmap0) refs
+      in (upd1, HM.insert ref0 upd2 hmap1)
+  where
+    handle_ref :: (updset, HM.HashMap ref updset) -> Cond ref
+      -> (updset, HM.HashMap ref updset)
+    handle_ref (upd_i0, hmap_i0) (Determined ref_i) =
+      let (upd_i1, hmap_i1) = solveReference lookup getRefs extUpds introCond (ref0:vis) hmap_i0 ref_i
+      in (extUpds upd_i0 upd_i1, hmap_i1)
+    handle_ref (upd_i0, hmap_i0) (Conditional cond ref_ia ref_ib) =
+      let (upd_ia, hmap_ia) = solveReference lookup getRefs extUpds introCond (ref0:vis) hmap_i0 ref_ia
+          (upd_ib, hmap_ib) = maybe (Nothing, hmap_ia)
+            (first Just . solveReference lookup getRefs extUpds introCond (ref0:vis) hmap_ia) ref_ib
+          upd_i1 = extUpds upd_i0 $ introCond $ Conditional cond upd_ia upd_ib
+      in (upd_i1, hmap_ib)
+
+-- TODO: This should work recursively
+-- | Lookup the field names and corresponding types, for a given record type
+-- constructor name.
+recTypConFields :: MonadEnv m ph
```

Comment:
`rec` is not the best prefix when both "record" and "recursion" are common words in your problem domain. 😄